### DISCUSSION TITLE
woodpile: SWC AST visitor in Javascript

### DISCUSSION THREAD

**Discussion by kwonoj** [swc-project/swc#7721] [Show and tell]
https://www.npmjs.com/package/woodpile

Woodpile is a utility library to traverse [SWC](https://github.com/swc-project/swc) ASTs in Javascript. It is a thin interop layer to the SWC's rust implementation of its visitor macro, attempt to provide consistent, synchronized mechanism to traverse ASTs in Javascript as well. This package could be useful to duck typing, prototyping SWC plugin in Javascript as its interface aims to provide similar experience as SWC's visitor macro.

For those reason, this package aims correctness over performance. There are inevitable costs to exchange data between Javascript to Rust, and vice versa. If you want to achieve peak performace, you should use SWC's visitor macro directly.

This is a userland alternative to handwritten JS visitor in @swc/core for the easier ast traversing.

### Usage

`visit` is a main interface to traverse AST.

Currently, `visit` accepts an object with `visit` property have corresponding callbacks to traverse.

```ts
const { visit } = require('woodpile');
const { parseSync } = require('@swc/core');

const ast = parseSync('console.log("Hello, World!")');

visit(ast, {
    visit: {
        // Callbacks with visit${NodeName} will be called recursively for the node
        visitProgram: (node, self) => {
            console.log('visitProgram', node);
        },
        visitExpr: (node) => {
            console.log('visitExpr', node);
        }
    },
});
```

**Comment by kwonoj** [comment]
Just to be clear, this is nothing related to writing-plugin-in-javascript.

**Comment by joshjhargreaves** [comment]
Hi there, I'm a bit confused as why you're recommending `swc/core/visitor` for performance  here but also look to be deprecating it: https://github.com/swc-project/swc/discussions/7673. Obviously I can very much see the benefit of using `woodpile` to prototype before writing rust plugins & `core/visitor` for that also. 

What is the status of `core/visitor` given its deprecated and what's the recomended path forward for JS interop in general?

**Reply by kwonoj** [reply]
The recommendation is `swc/core`, rust implementation of the visitor and not for the javascript one which isnplanned to deprecate.

The recommended path for js interop is writing a wasm plugin.

**Reply by joshjhargreaves** [reply]
Hi @kwonoj thanks for your reply. I tried your example wasm plugin for js interop I can't run the test example you have there. I left more info here: https://github.com/kwonoj/swc-plugin-js-transformer/issues/1. Thanks for time.
